pkg/controller/apicast/apicast_logic_reconciler.go

```diff
@@ -0,0 +1,423 @@
+		changed = true
+	}
+
+	return changed, nil
+}
+
+func (r *APIcastLogicReconciler) UserProvidedSecretResourceVersionAnnotations(userProvidedSecrets *apicastUserProvidedSecrets) map[string]string {
+	annotations := map[string]string{}
+
+	annotations[AdmPortalSecretResverAnnotation] = string(userProvidedSecrets.adminPortalCredentialsSecret.ResourceVersion)
+	return annotations
+}
+
+func (r *APIcastLogicReconciler) internalAPIcast(userProvidedSecrets *apicastUserProvidedSecrets) (apicast.APIcast, error) {
+	var err error
+
+	apicastFullName := "apicast-" + r.APIcastCR.Name
+	apicastExposedHost := apicast.ExposedHost{}
+	if r.APIcastCR.Spec.ExposedHost != nil {
+		apicastExposedHost.Host = r.APIcastCR.Spec.ExposedHost.Host
+		apicastExposedHost.TLS = r.APIcastCR.Spec.ExposedHost.TLS
+	}
+	apicastOwnerRef := asOwner(r.APIcastCR)
+
+	var deploymentEnvironment *string
+	if r.APIcastCR.Spec.DeploymentEnvironment != nil {
+		res := string(*r.APIcastCR.Spec.DeploymentEnvironment)
+		deploymentEnvironment = &res
+	}
+
+	var logLevel *string
+	if r.APIcastCR.Spec.LogLevel != nil {
+		res := string(*r.APIcastCR.Spec.LogLevel)
+		logLevel = &res
+	}
+
+	var managementAPIScope *string
+	if r.APIcastCR.Spec.ManagementAPIScope != nil {
+		res := string(*r.APIcastCR.Spec.ManagementAPIScope)
+		managementAPIScope = &res
+	}
+
+	deploymentAnnotations := r.UserProvidedSecretResourceVersionAnnotations(userProvidedSecrets)
+
+	apicastResult := apicast.APIcast{
+		DeploymentName:                   apicastFullName,
+		ServiceName:                      apicastFullName,
+		Replicas:                         int32(*r.APIcastCR.Spec.Replicas),
+		AppLabel:                         "apicast",
+		AdditionalAnnotations:            deploymentAnnotations,
+		ServiceAccountName:               *r.APIcastCR.Spec.ServiceAccount,
+		Image:                            *r.APIcastCR.Spec.Image,
+		ExposedHost:                      apicastExposedHost,
+		Namespace:                        r.APIcastCR.Namespace,
+		OwnerReference:                   &apicastOwnerRef,
+		AdminPortalCredentialsSecretName: userProvidedSecrets.adminPortalCredentialsSecret.Name,
+		DeploymentEnvironment:            deploymentEnvironment,
+		DNSResolverAddress:               r.APIcastCR.Spec.DNSResolverAddress,
+		EnabledServices:                  r.APIcastCR.Spec.EnabledServices,
+		ConfigurationLoadMode:            r.APIcastCR.Spec.ConfigurationLoadMode,
+		LogLevel:                         logLevel,
+		PathRoutingEnabled:               r.APIcastCR.Spec.PathRoutingEnabled,
+		ResponseCodesIncluded:            r.APIcastCR.Spec.ResponseCodesIncluded,
+		CacheConfigurationSeconds:        r.APIcastCR.Spec.CacheConfigurationSeconds,
+		ManagementAPIScope:               managementAPIScope,
+		OpenSSLPeerVerificationEnabled:   r.APIcastCR.Spec.OpenSSLPeerVerificationEnabled,
+	}
+
+	return apicastResult, err
+}
+
+func (r *APIcastLogicReconciler) namespacedName(object metav1.Object) types.NamespacedName {
```

Comment:
This method does not make sense as `APIcastLogicReconciler` method

Reply:
I put it there because currently it is used there. In case in the future we need it in another place we can refactor it if needed and think a new location for it.

Reply:
This kind of utility methods should always be defined externally.

The bad smell is very clear when you do not use the method receiver.

It's ok to leave there for now